src/backend/distributed/connection/connection_management.c

```diff
@@ -310,25 +311,86 @@ StartNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port,
 	if (!(flags & FORCE_NEW_CONNECTION))
 	{
 		/* check connection cache for a connection that's not already in use */
-		connection = FindAvailableConnection(entry->connections, flags);
+		MultiConnection *connection = FindAvailableConnection(entry->connections, flags);
 		if (connection)
 		{
 			return connection;
 		}
 	}
 
+
+	if (flags & NEVER_WAIT_FOR_CONNECTION)
+	{
+		/*
+		 * The caller doesn't want the connection manager to wait
+		 * until a connection slot is available on the remote node.
+		 * In the end, we might fail to establish connection to the
+		 * remote node as it might not have any space in
+		 * max_connections for this connection establishment.
+		 *
+		 * Still, we keep track of the connnection counter.
+		 */
+		IncrementSharedConnectionCounter(hostname, port);
+	}
+	else if (flags & OPTIONAL_CONNECTION)
+	{
+		/*
+		 * We can afford to skip establishing an optional connection. For
+		 * non-optional connections, we first retry for some time. If we still
+		 * cannot reserve the right to establish a connection, we prefer to
+		 * error out.
+		 */
+		if (!TryToIncrementSharedConnectionCounter(hostname, port))
+		{
+			return NULL;
+		}
+	}
+	else
+	{
+		WaitOrErrorForSharedConnection(hostname, port);
+	}
+
 	/*
 	 * Either no caching desired, or no pre-established, non-claimed,
 	 * connection present. Initiate connection establishment.
 	 */
+	volatile MultiConnection *connection = NULL;
+	PG_TRY();
+	{
+		connection = StartConnectionEstablishment(&key);
+	}
+	PG_CATCH();
+	{
+		/*
+		 * Something went wrong, make sure to decrement
+		 * here otherwise we'd leak the increment we have done for this
+		 * connection attempt.
+		 */
+		if (connection->pgConn == NULL)
+		{
+			DecrementSharedConnectionCounter(hostname, port);
+		}
+		else
+		{
+			CitusPQFinish((MultiConnection *) connection);
+		}
 
-	connection = StartConnectionEstablishment(&key);
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
 
-	dlist_push_tail(entry->connections, &connection->connectionNode);
 
-	ResetShardPlacementAssociation(connection);
+	/*
+	 * Cast volatile MultiConnection pointer to MultiConnection pointer
+	 * for ease of use.
+	 */
+	MultiConnection *newConnection = (MultiConnection *) connection;
 
-	return connection;
+	dlist_push_tail(entry->connections, &newConnection->connectionNode);
```

Comment:
I'm wondering whether we could move this to before StartConnectionEstablishment to avoid the PG_TRY. 

We obviously don't have a MultiConnection yet, but maybe we can create the struct here as well?